From deca5be1b3071974b4bd4f282a27e9a84b8de346 Mon Sep 17 00:00:00 2001
From: Hannes Winkler <hanneswinkler2000@web.de>
Date: Sun, 27 Jul 2025 22:28:44 +0200
Subject: [PATCH 10/10] Windows: Allow configuring VC++ compiler toolset
 version

The flutter engine currently does not compile with VC++ compiler
toolset 14.44,
as the bundled clang version is too old. (18 instead of the required
clang 19)

Allow explicitly configuring VC++ version so an older version can
be used,
via the `vs_version` gn argument.

By default, this gn argument is unset, and the build scripts will
still attempt
to use the latest VC++ version available, even if that is the
non-working
14.44 version. This is because explicitly looking for any concrete
version
would break some other working setups.

So the user has to manually specify `--gn-args=vc_version="14.42"`
/ `--gn-args="""vc_version=14.42"""` in escaped form for cmd,
when attempting to build the engine on a setup with the
non-working 14.44 installed.
---
 .../src/build/config/win/visual_studio_version.gni |  4 ++++
 engine/src/build/toolchain/win/setup_toolchain.py  | 14 ++++++++++----
 .../src/build/toolchain/win/win_toolchain_data.gni |  3 +++
 3 files changed, 17 insertions(+), 4 deletions(-)

diff --git a/engine/src/build/config/win/visual_studio_version.gni b/engine/src/build/config/win/visual_studio_version.gni
index fbb55b74dd5..2fc322808b4 100644
--- a/engine/src/build/config/win/visual_studio_version.gni
+++ b/engine/src/build/config/win/visual_studio_version.gni
@@ -20,6 +20,10 @@ declare_args() {
   # This value is the default location, override if you have a different
   # installation location.
   windows_sdk_path = "C:\Program Files (x86)\Windows Kits\8.1"
+
+  # The Version of the VC++ compiler toolset passed to
+  # vcvarsall.bat. Defaults to the latest version.
+  vc_version = ""
 }
 
 if (visual_studio_path == "") {
diff --git a/engine/src/build/toolchain/win/setup_toolchain.py b/engine/src/build/toolchain/win/setup_toolchain.py
index 521c24398ef..4a6ba6fe17c 100644
--- a/engine/src/build/toolchain/win/setup_toolchain.py
+++ b/engine/src/build/toolchain/win/setup_toolchain.py
@@ -101,7 +101,7 @@ def _LoadEnvFromBat(args):
   return variables.decode(errors='ignore')
 
 
-def _LoadToolchainEnv(cpu, toolchain_root, sdk_dir, target_store):
+def _LoadToolchainEnv(cpu, toolchain_root, sdk_dir, target_store, vc_version):
   """Returns a dictionary with environment variables that must be set while
   running binaries from the toolchain (e.g. INCLUDE and PATH for cl.exe)."""
   # Check if we are running in the SDK command line environment and use
@@ -186,6 +186,9 @@ def _LoadToolchainEnv(cpu, toolchain_root, sdk_dir, target_store):
     # building with a new and untested SDK. This should stay in sync with the
     # packaged toolchain in build/vs_toolchain.py.
     args.append(SDK_VERSION)
+
+    if vc_version is not None:
+      args.append('-vcvars_ver=%s' % (vc_version))
     variables = _LoadEnvFromBat(args)
   return _ExtractImportantEnvironment(variables)
 
@@ -225,11 +228,12 @@ def FindFileInEnvList(env, env_name, separator, file_name, optional=False):
 
 
 def main():
-  if len(sys.argv) != 7:
+  if len(sys.argv) != 7 and len(sys.argv) != 8:
     print('Usage setup_toolchain.py '
           '<visual studio path> <win sdk path> '
           '<runtime dirs> <target_os> <target_cpu> '
-          '<environment block name|none>')
+          '<environment block name|none> '
+          '[vc_version]')
     sys.exit(2)
   # toolchain_root and win_sdk_path are only read if the hermetic Windows
   # toolchain is set, that is if DEPOT_TOOLS_WIN_TOOLCHAIN is not set to 0.
@@ -245,6 +249,8 @@ def main():
   if (environment_block_name == 'none'):
     environment_block_name = ''
 
+  vc_version = sys.argv[7]
+
   if (target_os == 'winuwp'):
     target_store = True
   else:
@@ -271,7 +277,7 @@ def main():
   for cpu in cpus:
     if cpu == target_cpu:
       # Extract environment variables for subprocesses.
-      env = _LoadToolchainEnv(cpu, toolchain_root, win_sdk_path, target_store)
+      env = _LoadToolchainEnv(cpu, toolchain_root, win_sdk_path, target_store, vc_version)
       env['PATH'] = runtime_dirs + os.pathsep + env['PATH']
 
       vc_bin_dir = FindFileInEnvList(env, 'PATH', os.pathsep, 'cl.exe')
diff --git a/engine/src/build/toolchain/win/win_toolchain_data.gni b/engine/src/build/toolchain/win/win_toolchain_data.gni
index a6e8328b236..6a2534cab0d 100644
--- a/engine/src/build/toolchain/win/win_toolchain_data.gni
+++ b/engine/src/build/toolchain/win/win_toolchain_data.gni
@@ -14,6 +14,7 @@ declare_args() {
                     "win",
                     "x86",
                     "environment.x86",
+                    vc_version
                   ],
                   "scope")
 
@@ -26,6 +27,7 @@ declare_args() {
                     "win",
                     "x64",
                     "environment.x64",
+                    vc_version
                   ],
                   "scope")
   if (target_cpu == "arm64" || host_cpu == "arm64") {
@@ -38,6 +40,7 @@ declare_args() {
                       "win",
                       "arm64",
                       "environment.arm64",
+                      vc_version
                     ],
                     "scope")
   }
-- 
2.50.1.windows.1

